File: naddu/naddu_2/src/qsearch.js
function qsearch(ply, alpha, beta) {

  if (ply === MAX_PLY - 1)
    return evaluate(nodes[ply]);

  const tc = timeControl;

  tcCheck();
  if (tc.finished)
    return 0;

  tc.nodes++;

  const node = nodes[ply];
  const pos = node.pos;
  
  // check tt
  const ttIndex = ttGet(pos);
  if (ttIndex >= 0) {
    const score = ttScoreFromTT(ttGetScore(ttIndex), ply);
    const type = ttGetType(ttIndex);
    if (type === TT_EXACT || (type === TT_BETA && score >= beta) || (type === TT_ALPHA && score <= alpha)) {
      return score;
    }
  }

  // stand pat ?
  const standPat = evaluate(node);
  if (node.draw)
    return 0;
  if (standPat >= beta)
    return standPat;
  if (standPat > alpha)
    alpha = standPat;

  const nextNode = nodes[ply + 1];
  const nextPos = nextNode.pos;
  const stmi = pos.stm >> 3;

  const ttMove = ttIndex >= 0 && (ttGetMove(ttIndex) & MOVE_FLAG_CAPTURE) ? moveIsProbablyLegal(node, ttGetMove(ttIndex)) : 0;

  initNextMoveQS(node, ttMove);

  let move;

  while ((move = getNextMoveQS(node))) {

    posSet(nextPos, pos);
    makeMove(move, nextPos);

    if (isAttacked(nextPos, nextPos.kings[stmi], nextPos.stm))
      continue;

    const score = -qsearch(ply + 1, -beta, -alpha);

    if (tc.finished)
      return 0;

    if (score >= beta)
      return score;

    if (score > alpha)
      alpha = score;
  }

  return alpha;

}

File: naddu/naddu_2/c/qsearch.c
#include "qsearch.h"
#include "node.h"
#include "pos.h"
#include "eval.h"
#include "tt.h"
#include "tc.h"
#include "attack.h"
#include "make.h"
#include "move.h"
#include "iterate.h"

int qsearch(int ply, int alpha, int beta) {
  if (ply == MAX_PLY - 1)
    return evaluate(&nodes[ply]);

  tc_check();
  if (tc.finished)
    return 0;

  tc.nodes++;

  Node *node = &nodes[ply];
  Pos *pos = &node->pos;

  int tt_index = tt_get(pos);
  if (tt_index >= 0) {
    int score = tt_score_from_tt(tt_get_score(tt_index), ply);
    int type = tt_get_type(tt_index);
    if (type == TT_EXACT || (type == TT_BETA && score >= beta) || (type == TT_ALPHA && score <= alpha)) {
      return score;
    }
  }

  int stand_pat = evaluate(node);
  if (node->draw)
    return 0;
  if (stand_pat >= beta)
    return stand_pat;
  if (stand_pat > alpha)
    alpha = stand_pat;

  Node *next_node = &nodes[ply + 1];
  Pos *next_pos = &next_node->pos;
  int stmi = pos->stm >> 3;

  uint32_t tt_move = 0;
  if (tt_index >= 0) {
    uint32_t tm = tt_get_move(tt_index);
    if (tm & MOVE_FLAG_CAPTURE) {
      tt_move = move_is_probably_legal(node, tm);
    }
  }

  init_next_move_qs(node, tt_move);

  uint32_t move;

  while ((move = get_next_move_qs(node))) {
    pos_copy(next_pos, pos);
    make_move(move, next_pos);

    if (is_attacked(next_pos, next_pos->kings[stmi], next_pos->stm))
      continue;

    int score = -qsearch(ply + 1, -beta, -alpha);

    if (tc.finished)
      return 0;

    if (score >= beta)
      return score;

    if (score > alpha)
      alpha = score;
  }

  return alpha;
}
